File: backend/module-ai/src/main/java/com/portfolio/module/ai/service/LlmService.java
package com.portfolio.module.ai.service;

import com.portfolio.module.ai.config.AiProperties;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

import java.time.Duration;
import java.util.List;
import java.util.Map;

@Slf4j
@Service
public class LlmService {

    private static final Duration OLLAMA_TIMEOUT = Duration.ofMinutes(5);
    private static final Duration GEMINI_TIMEOUT = Duration.ofMinutes(2);
    private static final String GEMINI_API_BASE = "https://generativelanguage.googleapis.com/v1beta";

    private final AiProperties aiProperties;
    private final WebClient ollamaWebClient;
    private final WebClient geminiWebClient;

    public LlmService(
            @Qualifier("aiWebClientBuilder") WebClient.Builder webClientBuilder,
            AiProperties aiProperties) {
        this.aiProperties = aiProperties;
        this.ollamaWebClient = webClientBuilder.clone()
                .baseUrl(aiProperties.getOllamaHost())
                .build();
        this.geminiWebClient = webClientBuilder.clone()
                .baseUrl(GEMINI_API_BASE)
                .build();
    }

    /**
     * Notion 콘텐츠를 기반으로 블로그 초안을 생성한다.
     * ai.provider 설정에 따라 Ollama 또는 Gemini를 사용한다.
     */
    public Mono<String> generateDraft(String notionContent, String instructions) {
        String prompt = buildPrompt(notionContent, instructions);
        String provider = aiProperties.getProvider();

        log.info("LLM 초안 생성 시작 (provider: {})", provider);

        return switch (provider.toLowerCase()) {
            case "gemini" -> generateWithGemini(prompt);
            case "ollama" -> generateWithOllama(prompt);
            default -> {
                log.warn("알 수 없는 provider '{}', 기본값 ollama로 폴백", provider);
                yield generateWithOllama(prompt);
            }
        };
    }

    public String getProvider() {
        return aiProperties.getProvider();
    }

    @SuppressWarnings("unchecked")
    private Mono<String> generateWithOllama(String prompt) {
        Map<String, Object> requestBody = Map.of(
                "model", aiProperties.getOllamaModel(),
                "prompt", prompt,
                "stream", false,
                "options", Map.of(
                        "num_predict", 2048,
                        "temperature", 0.7
                )
        );

        return ollamaWebClient.post()
                .uri("/api/generate")
                .bodyValue(requestBody)
                .retrieve()
                .bodyToMono(new ParameterizedTypeReference<Map<String, Object>>() {})
                .timeout(OLLAMA_TIMEOUT)
                .map(response -> {
                    String result = (String) response.get("response");
                    log.info("Ollama 초안 생성 완료 ({}자)", result != null ? result.length() : 0);
                    return result != null ? result : "";
                })
                .doOnError(e -> log.error("Ollama 호출 실패: {}", e.getMessage()));
    }

    @SuppressWarnings("unchecked")
    private Mono<String> generateWithGemini(String prompt) {
        Map<String, Object> requestBody = Map.of(
                "contents", List.of(
                        Map.of("parts", List.of(
                                Map.of("text", prompt)
                        ))
                )
        );

        return geminiWebClient.post()
                .uri("/models/{model}:generateContent?key={apiKey}",
                        aiProperties.getGeminiModel(),
                        aiProperties.getGeminiApiKey())
                .bodyValue(requestBody)
                .retrieve()
                .bodyToMono(new ParameterizedTypeReference<Map<String, Object>>() {})
                .timeout(GEMINI_TIMEOUT)
                .map(response -> {
                    try {
                        List<Map<String, Object>> candidates =
                                (List<Map<String, Object>>) response.get("candidates");
                        if (candidates == null || candidates.isEmpty()) return "";

                        Map<String, Object> content =
                                (Map<String, Object>) candidates.get(0).get("content");
                        if (content == null) return "";

                        List<Map<String, Object>> parts =
                                (List<Map<String, Object>>) content.get("parts");
                        if (parts == null || parts.isEmpty()) return "";

                        String result = (String) parts.get(0).get("text");
                        log.info("Gemini 초안 생성 완료 ({}자)", result != null ? result.length() : 0);
                        return result != null ? result : "";
                    } catch (Exception e) {
                        log.error("Gemini 응답 파싱 실패: {}", e.getMessage());
                        return "";
                    }
                })
                .doOnError(e -> log.error("Gemini 호출 실패: {}", e.getMessage()));
    }

    private String buildPrompt(String notionContent, String instructions) {
        StringBuilder sb = new StringBuilder();
        sb.append("You are a professional blog post writer. ");
        sb.append("Based on the following Notion page content, generate a well-structured blog post draft.\n\n");

        if (instructions != null && !instructions.isBlank()) {
            sb.append("## Additional Instructions\n");
            sb.append(instructions).append("\n\n");
        }

        sb.append("## Source Content (from Notion)\n");
        sb.append(notionContent).append("\n\n");
        sb.append("## Requirements\n");
        sb.append("- Write in the same language as the source content\n");
        sb.append("- Use clear headings and structure\n");
        sb.append("- Keep the technical accuracy of the original content\n");
        sb.append("- Make it engaging and readable\n");
        sb.append("- Output in Markdown format\n");

        return sb.toString();
    }
}

File: ai-backend/app/llm_service.py
import time
import logging

from langchain_core.language_models import BaseChatModel
from langchain_google_genai import ChatGoogleGenerativeAI
from langchain_ollama import ChatOllama

from app.config import settings

logger = logging.getLogger(__name__)


def _build_model() -> BaseChatModel:
    if settings.ai_provider.lower() == "gemini":
        return ChatGoogleGenerativeAI(
            model=settings.gemini_model,
            google_api_key=settings.gemini_api_key,
            temperature=0.3,
            max_output_tokens=256,
            timeout=30,
        )
    return ChatOllama(
        base_url=settings.ollama_host,
        model=settings.ollama_model,
        temperature=0.3,
        timeout=60,
    )


def summarize(content: str, title: str, max_length: int) -> tuple[str, int]:
    """텍스트 요약. (summary, duration_ms) 반환."""
    model = _build_model()

    prompt = (
        f"다음 블로그 글의 핵심 내용을 한국어로 {max_length}자 이내로 요약해주세요.\n"
        f"- 요약문만 출력하세요 (부가 설명, 따옴표, 접두어 없이)\n"
        f"- 독자가 글의 핵심을 빠르게 파악할 수 있도록 작성하세요\n\n"
    )
    if title:
        prompt += f"제목: {title}\n\n"
    prompt += f"본문:\n{content}"

    start = time.time()
    result = model.invoke(prompt).content.strip()
    duration_ms = int((time.time() - start) * 1000)

    # 따옴표 제거
    if result.startswith('"') and result.endswith('"'):
        result = result[1:-1]

    # 길이 제한
    if len(result) > max_length:
        result = result[:max_length]

    logger.info(
        "Summarization completed: provider=%s, model=%s, durationMs=%d, length=%d",
        settings.ai_provider,
        settings.gemini_model if settings.ai_provider.lower() == "gemini" else settings.ollama_model,
        duration_ms,
        len(result),
    )
    return result, duration_ms


def generate_draft(notion_content: str, instructions: str) -> tuple[str, int]:
    """Notion 콘텐츠 기반 블로그 초안 생성. (content, duration_ms) 반환."""
    model = _build_model()

    prompt = (
        "You are a professional blog post writer. "
        "Based on the following Notion page content, generate a well-structured blog post draft.\n\n"
    )
    if instructions:
        prompt += f"## Additional Instructions\n{instructions}\n\n"
    prompt += (
        f"## Source Content (from Notion)\n{notion_content}\n\n"
        "## Requirements\n"
        "- Write in the same language as the source content\n"
        "- Use clear headings and structure\n"
        "- Keep the technical accuracy of the original content\n"
        "- Make it engaging and readable\n"
        "- Output in Markdown format\n"
    )

    start = time.time()
    result = model.invoke(prompt).content
    duration_ms = int((time.time() - start) * 1000)

    logger.info("Draft generation completed: provider=%s, durationMs=%d, length=%d",
                settings.ai_provider, duration_ms, len(result))
    return result, duration_ms
